src/ucs/datastruct/hlist.h

```diff
@@ -165,6 +165,57 @@ ucs_hlist_extract_head(ucs_hlist_head_t *head)
 }
 
 
+/**
+ * Get list head element as the containing type, assuming the list is not empty.
+ *
+ * @param _head    List head.
+ * @param _type    Containing structure type.
+ * @param _member  List element inside the containing structure.
+ *
+ * @note If the list is empty this macro has undefined behavior.
+ */
+#define ucs_hlist_head_elem(_head, _type, _member) \
+    ucs_container_of((_head)->ptr, _type, _member)
+
+
+/**
+ * Get list next element as the containing type.
+ *
+ * @param _elem    List element.
+ * @param _type    Containing structure type.
+ * @param _member  List element inside the containing structure.
+ */
+#define ucs_hlist_next_elem(_elem, _member) \
+    ucs_container_of(ucs_list_next(&(_elem)->_member.list, ucs_hlist_link_t, \
+                                   list), \
+                     typeof(*_elem), _member)
+
+
+/**
+ * Iterate over detached-head list.
+ *
+ * @param _elem     Variable to hold the current list element
+ * @param _head     Pointer to list head.
+ * @param _member   List element inside the containing structure.
+ *
+ * @note The iteration is implemented by first setting the element to NULL, then
+ * inside 'for' loop condition (which is done before each iteration), we advance
+ * the element pointer and check for end condition: in the first iteration,
+ * when elem is NULL, we check that the list is not empty. For subsequent
+ * iterations, we check that elem has not reached the list head yet.
+ */
+#define ucs_hlist_for_each(_elem, _head, _member) \
```

Comment:
@yosefe I thought that `head::list` container doesn't contain `head::ptr`. so, 
```
             ((_elem = ucs_hlist_next_elem(_elem, _member)) != \
                     ucs_hlist_head_elem(_head, typeof(*_elem), _member)); \
```
condition is always true. and `(_elem == NULL)` + `((_elem = ucs_hlist_head_elem(_head, typeof(*_elem), _member)) !=
                     ucs_container_of(NULL, typeof(*(_elem)), _member))` is loop termination condition

if  so, it can be re-written:
```
#define ucs_hlist_for_each(_elem, _head, _member) \
    for (_elem = ucs_hlist_head_elem(_head, typeof(*_elem), _member);
         _elem != ucs_container_of(NULL, typeof(*(_elem)), _member); \
         _elem = ucs_hlist_next_elem(_elem, _member))
```

Reply:
the list is circular, so if it's not empty, the condition `_elem != ucs_container_of(NULL, typeof(*(_elem)), _member);` will always be true, which will result in infinite loop.
`(_elem == NULL)` + `((_elem = ucs_hlist_head_elem(_head, typeof(*_elem), _member)) !=                    ucs_container_of(NULL, typeof(*(_elem)), _member))` is just for the case of empty list.

Reply:
Sorry, yes, you're right
thank you for explanation